codalab/lib/bundle_fuse.py

```diff
@@ -0,0 +1,162 @@
+#!/usr/bin/env python
+
+from __future__ import with_statement
+
+import os
+import sys
+import errno
+import stat
+import time
+
+from contextlib import closing
+
+from codalab.lib.fuse import FUSE, FuseOSError, Operations
+from codalab.client.json_api_client import JsonApiRelationship
+from codalab.lib.path_util import normalize
+
+class MWT(object):
+    """
+    Memoize With Timeout
+    https://code.activestate.com/recipes/325905-memoize-decorator-with-timeout/
+    """
+
+    _caches = {}
+    _timeouts = {}
+
+    def __init__(self,timeout=2):
+        self.timeout = timeout
+
+    def collect(self):
+        """Clear cache of results which have timed out"""
+        for func in self._caches:
+            cache = {}
+            for key in self._caches[func]:
+                if (time.time() - self._caches[func][key][1]) < self._timeouts[func]:
+                    cache[key] = self._caches[func][key]
+            self._caches[func] = cache
+
+    def __call__(self, f):
+        self.cache = self._caches[f] = {}
+        self._timeouts[f] = self.timeout
+
+        def func(*args, **kwargs):
+            kw = kwargs.items()
+            kw.sort()
+            key = (args, tuple(kw))
+            try:
+                v = self.cache[key]
+                if (time.time() - v[1]) > self.timeout:
+                    raise KeyError
+            except KeyError:
+                v = self.cache[key] = f(*args,**kwargs),time.time()
+            return v[0]
+        func.func_name = f.func_name
+
+        return func
+
+class Memoize(MWT):
+    '''A superset of MWT that adds the possibility to yank paths from cached results'''
+    def yank_path(self, path):
+        """Clear cache of results from a specific path"""
+        for func in self._caches:
+            cache = {}
+            for key in self._caches[func].keys():
+                if path in key[0]:
+                    del self._caches[func][key]
+
+class BundleFuse(Operations):
+    def __init__(self, client, target):
+        self.client = client
+        self.target = target
+        self.bundle_uuid = target[0]
+        self.fd = 0 # file descriptor
+
+    # Helpers
+    # =======
+
+    @Memoize(timeout=5)
+    def _get_info(self, path):
+        ''' Set a request through the json api client to get info about the bundle '''
+        info = self.client.fetch_contents_info(self.bundle_uuid, path, 1)
```

Comment:
When a file doesn't exist, `info` will be `None`.

(In the future, a `NotFoundError` will be thrown instead, but I can handle that change.)